evm/src/cpu/kernel/asm/sha2/ops.asm

```diff
@@ -0,0 +1,189 @@
+// u32 addition (discarding 2^32 bit)
+%macro add_u32
+    // stack: x, y
+    ADD
+    // stack: x + y
+    DUP1
+    // stack: x + y, x + y
+    %shr_const(32)
+    // stack: (x + y) >> 32, x + y
+    %shl_const(32)
+    // stack: ((x + y) >> 32) << 32, x + y
+    SWAP1
+    // stack: x + y, ((x + y) >> 32) << 32
+    SUB
+    // stack: x + y - ((x + y) >> 32) << 32
+%endmacro
+
+
+// 32-bit right rotation
+%macro rotr
+    // stack: rot, value
+    DUP2
+    DUP2
+    // stack: rot, value, rot, value
+    SHR
+    // stack: value >> rot, rot, value
+    %stack (shifted, rot, value) -> (rot, value, shifted)
+    // stack: rot, value, value >> rot
+    PUSH 32
+    SUB
+    // stack: 32 - rot, value, value >> rot
+    SHL
+    // stack: value << (32 - rot), value >> rot
+    PUSH 32
+    PUSH 1
+    SWAP1
+    SHL
+    // stack: 1 << 32, value << (32 - rot), value >> rot
+    SWAP1
+    MOD
+    // stack: (value << (32 - rot)) % (1 << 32), value >> rot
+    ADD
+%endmacro
+
+// 32-bit left rotation
+%macro rotl
+    // stack: rot, value
+    DUP2
+    DUP2
+    // stack: rot, value, rot, value
+    PUSH 32
+    SUB
+    // stack: 32 - rot, value, rot, value
+    SHR
+    // stack: value >> (32 - rot), rot, value
+    %stack (shifted, rot, value) -> (rot, value, shifted)
+    // stack: rot, value, value >> (32 - rot)
+    SHL
+    // stack: value << rot, value >> (32 - rot)
+    PUSH 32
+    PUSH 1
+    SWAP1
+    SHL
+    // stack: 1 << 32, value << rot, value >> (32 - rot)
+    SWAP1
+    MOD
```

Comment:
I'd suggest adding a `truncate_to_u32` macro (it can be e.g. `%and_const(0xFFFFFFFF)`), and then we can use it here in place of this `% (1 << 32)` code